crates/primitives/src/sealed.rs

```diff
@@ -0,0 +1,68 @@
+use crate::B256;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+/// A consensus hashable item, with its memoized hash.
+///
+/// We do not implement
+pub struct Sealed<T> {
+    /// The inner item
+    inner: T,
+    /// Its hash.
+    seal: B256,
+}
+
+impl<T> core::ops::Deref for Sealed<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        self.inner()
+    }
+}
+
+impl<T> Sealed<T> {
+    /// Instantiate without performing the hash. This should be used carefully.
+    pub const fn new_unchecked(inner: T, seal: B256) -> Self {
+        Self { inner, seal }
+    }
+
+    /// Decompose into parts.
+    #[allow(clippy::missing_const_for_fn)] // false positive
+    pub fn into_parts(self) -> (T, B256) {
+        (self.inner, self.seal)
+    }
+
+    /// Get the inner item.
+    #[inline(always)]
+    pub const fn inner(&self) -> &T {
+        &self.inner
+    }
+
+    /// Get the hash.
+    #[inline(always)]
+    pub const fn seal(&self) -> B256 {
+        self.seal
+    }
+
+    /// Geth the hash (alias for [`Self::seal`]).
+    #[inline(always)]
+    pub const fn hash(&self) -> B256 {
+        self.seal()
+    }
+}
+
+/// Sealeable objects.
+pub trait Sealable: Sized {
+    /// Calculate the seal hash, this may be slow.
+    fn hash(&self) -> B256;
```

Comment:
hmm, should this be hash_slow?

perhaps this could clash with `Sealed::hash`
`<Sealed Deref::Target>::hash`

Reply:
easy 👍